### PR TITLE
Login UI

### PR DESCRIPTION
**Pull request by BanFilip**
Implemented Login UI  (Sorry for new pods)

Simulator Screen Shot - iPhone 12 - 2021-11-10 at 22 12 13

Simulator Screen Shot - iPhone 12 - 2021-11-10 at 22 12 05

Simulator Screen Shot - iPhone 12 - 2021-11-10 at 22 12 02

### CODE REVIEW

File: iOS-TvShows/Podfile [hunk 1]
```diff
@@ -11,4 +11,6 @@ target 'iOS-TvShows' do
   pod 'Alamofire'
   pod 'Kingfisher'
   pod 'KeychainAccess'
+  pod 'SnapKit'
```
Comment (fgulan) [inline iOS-TvShows/Podfile:14]: Few things to comment here:

1. Pod installation should always be a separate commit without any changes on your code side. There are few reasons for that: when reviewing PR it is much easier to exclude that commit and solely review your code. Also it is easier to cherry pick some future if needed. And commit is not atomic
2. Even though I told you it is fine to use anything for keyboard/constraints that doesn't include using Pods 😄 we really try to use as less as possible pods, especially UI related ones - they tend to have issues with maintaining + complexity for a new person on project to learn how to use exact that pod.

Comment (fgulan) [reply iOS-TvShows/Podfile:14]: For example, keyboard can be handled with this approach + Rx:

```swift
import UIKit
import RxSwift
import RxCocoa

public class KeyboardHandler {

    private init() {}

    public struct PresentingParams {
        let keyboardSize: CGSize
        let animationDuration: Double
        let state: State

        enum State {
            case show
            case hide
        }
    }

    public enum Insets {
        case forShown(inset: CGFloat)
        case forHidden(inset: CGFloat)
        case always(inset: CGFloat)
        case automatic
        case never
    }

}

public extension KeyboardHandler {

    static func register(
        handler: @escaping (_ params: PresentingParams) -> Void
    ) -> Disposable {
        return keyboardPresenting
            .observeOn(MainScheduler.instance)
            .subscribe(onNext: { handler($0) })
    }

    static func register(
        animatedHandler: @escaping (_ params: PresentingParams) -> Void
    ) -> Disposable {
        return KeyboardHandler
            .register(handler: { (params) in
                UIView.animate(withDuration: params.animationDuration) {
                    animatedHandler(params)
                }
            })
    }

}

public extension KeyboardHandler {

    static func register(
        scrollView: UIScrollView,
        additionalInsets: Insets = .never
    ) -> Disposable {
        return KeyboardHandler
            .register(handler: {
                let height = additionalInsets.add(to: $0.keyboardSize.height, with: $0.state)
                scrollView.contentInset.bottom = height
                scrollView.scrollIndicatorInsets.bottom = height
            })
    }

    static func register(
        constraint: NSLayoutConstraint,
        additionalInsets: Insets = .never,
        aditionalAnimationBlock: (() -> Void)? = nil
    ) -> Disposable {
        return KeyboardHandler
            .register(animatedHandler: {
                let height = additionalInsets.add(to: $0.keyboardSize.height, with: $0.state)
                constraint.constant = height
                aditionalAnimationBlock?()
            })
    }

}

public extension KeyboardHandler {

    static private(set) var keyboardPresenting: Observable<PresentingParams> = {
        let center = NotificationCenter.default

        let show = center.rx
            .notification(UIResponder.keyboardWillShowNotification)
            .map { PresentingParams(state: .show, notification: $0) }

        let hide = center.rx
            .notification(UIResponder.keyboardWillHideNotification)
            .map { PresentingParams(state: .hide, notification: $0) }

        return Observable.merge([show, hide])
            .share(replay: 1, scope: .forever)
            .unwrap()
    }()

}

private extension KeyboardHandler.Insets {

    func add(
        to value: CGFloat,
        with state: KeyboardHandler.PresentingParams.State
    ) -> CGFloat {
        switch (self, state) {
        case (.always(let inset), _):
            return value + inset
        case (.forShown(let inset), .show):
            return value + inset
        case (.forHidden(let inset), .hide):
            return value + inset
        case (.automatic, .show):
            return value - UIView.bottomSafeAreaInsets
        default:
            return value
        }
    }

}

private extension KeyboardHandler.PresentingParams {

    init?(state: State, notification: Notification) {
        guard
            let animationDuration = notification.get(
                Double.self,
                for: UIResponder.keyboardAnimationDurationUserInfoKey
            )
            else { return nil }

        self.keyboardSize = KeyboardHandler.PresentingParams._size(for: state, from: notification)
        self.state = state
        self.animationDuration = animationDuration
    }

    private static func _size(
        for state: State,
        from notification: Notification
    ) -> CGSize {
        guard state != .hide else {
            return .zero
        }

        guard
            let rect = notification.get(
                NSValue.self,
                for: UIResponder.keyboardFrameEndUserInfoKey
                )?.cgRectValue
            else {
                return .zero
        }
        return rect.size
    }

}

```

Comment (fgulan) [reply iOS-TvShows/Podfile:14]: While for constraints we are using NutsAndBolts + anchors - we've used snap kit but lots of magic is going behind it while anchors are really easy to use and giving the developer a better grasp what's happening.

Comment (fgulan) [reply iOS-TvShows/Podfile:14]: I mean you can leave SnapKit but we won't be using it on the projects 😄

File: iOS-TvShows/iOS-TvShows/Application/AppDelegate.swift [hunk 2]
```diff
@@ -25,4 +26,8 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
         window.makeKeyAndVisible()
         return true
     }
+
+    private func beforeAppUIWillLoad() {
+        UIFont.loadAppFonts()
```
Comment (fgulan) [inline iOS-TvShows/iOS-TvShows/Application/AppDelegate.swift]: Hmm what's the purpose behind this code in case you are using system fonts?

File: iOS-TvShows/iOS-TvShows/Common/Extensions/UIColor+Extensions.swift [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+//  iOS-TvShows
+//
+//  Created by Filip Ban on 09.11.2021..
+//
+
+import UIKit
+
+extension UIColor {
+
+    static var appWhite: UIColor {
```
Comment (fgulan) [inline iOS-TvShows/iOS-TvShows/Common/Extensions/UIColor+Extensions.swift]: This is a good approach for dealing with colors - aka grouping them on a single place. But rather than hardcoding hex values in code, we are using Color xcassets where you can preview the color and access it in the code via name

File: iOS-TvShows/iOS-TvShows/Common/UI/Bundles/BundleImage.swift [hunk 4]
```diff
@@ -0,0 +1,17 @@
+//
+//  BundleImage.swift
+//  iOS-TvShows
+//
+//  Created by Filip Ban on 09.11.2021..
+//
+
+enum BundleImage: String {
```
Comment (fgulan) [inline iOS-TvShows/iOS-TvShows/Common/UI/Bundles/BundleImage.swift]: Rather than having manual entries for names we have a tool which will autogenerate this from XCAssets folder ;) https://github.com/infinum/redbreast

File: iOS-TvShows/iOS-TvShows/Common/Extensions/UIViewController+Extensions.swift [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+    var leftBarButton: UIBarButtonItem? {
+        navigationItem.leftBarButtonItem
+    }
+
+    var rightBarButton: UIBarButtonItem? {
+        navigationItem.rightBarButtonItem
+    }
+
+    func hideNavigationBar() {
```
Comment (fgulan) [inline iOS-TvShows/iOS-TvShows/Common/Extensions/UIViewController+Extensions.swift]: Please pass the animated flag here since this method should only be used from `viewWillAppear` otherwise you'll have weird glitches on UI. Also even tho we tend to have less code to have better readability, think about onboarding new colleague on project of yours where you are using this approach - it's a bit non-iOS to have hideNavigationBar() on view controller since all iOS developers know they should hide it on navigation controller. Having these kinds of method short your code for few character but increase the complexity of understanding your code for common iOS devs

File: iOS-TvShows/iOS-TvShows/Common/Services/AppKeyboardService/KeyboardViewController.swift [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+    var bottomConstraint: Constraint?
+
+    weak var appKeyboardService = AppKeyboardService.instance
+
+    convenience init(appKeyboardService: AppKeyboardServiceInterface) {
+        self.init()
+    }
+
+    override func viewWillAppear(_ animated: Bool) {
```
Comment (fgulan) [inline iOS-TvShows/iOS-TvShows/Common/Services/AppKeyboardService/KeyboardViewController.swift]: You are missing super call here.

File: iOS-TvShows/iOS-TvShows/Common/Services/AppKeyboardService/KeyboardViewController.swift [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+//  KeyboardViewController.swift
+//  iOS-TvShows
+//
+//  Created by Filip Ban on 10.11.2021..
+//
+
+import UIKit
+import SnapKit
+
+class KeyboardViewController: UIViewController, KeyboardSubscriberInterface {
```
Comment (fgulan) [inline iOS-TvShows/iOS-TvShows/Common/Services/AppKeyboardService/KeyboardViewController.swift]: While it is cool to have shared logic between view controllers for handling keyboard, but composition is a better approach rather than subclassing. Sometimes you'll have the case where your view controller has a specific subclass and you won't be able to add subclass for your keyboard view controller.

```swift
protocol KeyboardViewControllerSubscriberInterface: KeyboardSubscriberInterface {
    var bottomConstraint: Constraint? { get }
}

extension KeyboardViewControllerSubscriberInterface where Self: UIViewController {

    func onKeyboardFrameChanged(
        endFrame: CGRect,
        endFrameY: CGFloat,
        animationDuration: Double,
        bottomPadding: CGFloat
    ) {
        if endFrameY >= UIScreen.main.bounds.size.height {
            bottomConstraint?.update(offset: 0)
        } else {
            let frameHeight = endFrame.size.height
            bottomConstraint?.update(offset: bottomPadding - frameHeight)
        }

        UIView.animate(withDuration: animationDuration) { [weak self] in
            self?.view.layoutIfNeeded()
        }
    }
}

class SomeViewController: KeyboardViewControllerSubscriberInterface {

    var bottomConstraint: Constraint? { view.someConstraint ... }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(...)
        appKeyboardService?.subscribe(subscriber: self)
    }
}
```

Comment (fgulan) [reply iOS-TvShows/iOS-TvShows/Common/Services/AppKeyboardService/KeyboardViewController.swift]: Also, what about scrollviews where you don't want to change bottom constraint, rather content inset?

File: iOS-TvShows/iOS-TvShows/Common/UI/Bundles/BundleFont.swift [hunk 8]
```diff
@@ -0,0 +1,15 @@
+//
+//  BundleFont.swift
+//  iOS-TvShows
+//
+//  Created by Filip Ban on 09.11.2021..
+//
+
+enum BundleFont: String, CaseIterable {
+
+    case regular = "SFProDisplay-Regular"
```
Comment (fgulan) [inline iOS-TvShows/iOS-TvShows/Common/UI/Bundles/BundleFont.swift]: Hmm why did you add these fonts? Those are the `UIFont.system`, no need for custom adding them 😄

**Review by fgulan — COMMENTED** [review]
It is a bit hard to review this big chunk via PR, let's arrange some meet and go live over it 😉

File: iOS-TvShows/iOS-TvShows/Common/UI/Bundles/BundleFont.swift (hunk 8 above)
Comment (BanFilip) [reply iOS-TvShows/iOS-TvShows/Common/UI/Bundles/BundleFont.swift]: Didnt know :D
